File: iam-policy-lens/scripts/python/analyzer.py
"""
Python Cloud Access Scanner
===========================

Problem Statement:
------------------
This script performs static analysis on a Python codebase to identify Google Cloud, Vertex AI, and Google GenAI client library invocations.
The primary goal is to locate and extract these high-level, Pythonic method calls (e.g., `google.cloud.storage.Client.create_bucket`).

How It Works:
-------------
1. GAPIC Call Extraction (gapic.py): 
    The scanner uses `mypy.build` to generate an Abstract Syntax Tree (AST) of the target project, which handles tricky type-resolution paths.
    It walks the AST looking for `CallExpr` nodes and traces back nested `MemberExpr` chains to resolve fully qualified import names.

How to Run:
-----------
Run the script passing the target repository source path as the first argument:
    .venv/bin/python3 scripts/python/analyzer.py <path_to_project>
    .venv/bin/python3  scripts/python/analyzer.py /Users/petrusca/Google/adk-samples/python/agents/data-science
    .venv/bin/python3 scripts/python/analyzer.py /Users/petrusca/Google/gcp_python_iam_analyzer/gcp_cost_optimizer_agent
"""
import os
import sys
import time

sys.path.append(os.path.dirname(os.path.abspath(__file__)))

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage:")
        print("  python scripts/python/analyzer.py <path_to_project> [python_env_path]")
        sys.exit(1)
        
    project_path = sys.argv[1]
    python_env = sys.argv[2] if len(sys.argv) >= 3 else None
    
    print(f"Scanning: {project_path} for GAPIC calls")
    if python_env:
        print(f"Using Python environment: {python_env}")
        
    start_time = time.time()
    import scanner
    raw_calls = scanner.find_gapic_calls(project_path, python_env)
    elapsed_time = time.time() - start_time
    
    if raw_calls:
        from collections import defaultdict
        grouped_calls = defaultdict(list)
        for call in raw_calls:
            full_path = os.path.abspath(call.file_path)
            grouped_calls[full_path].append(call)
            
        for full_path, calls in sorted(grouped_calls.items()):
            print(f"\n📄 File: {full_path}")
            rel_path = os.path.relpath(full_path, project_path) if os.path.isabs(full_path) else full_path
            for i, call in enumerate(sorted(calls, key=lambda x: x.line)):
                if i > 0:
                    print()
                print(f"     {rel_path}:{call.line}: `{call.source_line}`")
                print(f"     Method: {call.fullname} [{call.resolution}]")
                if getattr(call, 'client_fullname', None):
                    print(f"     Client: {call.client_fullname}")
                if getattr(call, 'credentials', None):
                    print(f"     Credentials: {call.credentials.source}")
                    print(f"     Provenance: {call.credentials.provenance}")
                    print(f"     Identity: {call.credentials.identity}")
                    
                from permissions import gapic2permission
                permissions = gapic2permission(call.fullname)
                if permissions:
                    print(f"     Permissions: {permissions}")
                    
        # Generate and print GCP IAM V3 Allow Policies
        from policy import generate_iam_policies
        import json
        
        print("\n====================================================")
        print("🔒 Generated GCP IAM V3 Allow Policies")
        print("====================================================")
        
        sa_email = os.getenv("GCP_SERVICE_ACCOUNT")
        policies = generate_iam_policies(raw_calls, sa_email)
        for p in policies:
            print(f"\n📍 Attachment Point: {p['attachment_point']}")
            print(json.dumps(p['policy'], indent=4))
    else:
        print("No relevant GAPIC calls found.")
        
    print(f"\nScan completed in {elapsed_time:.2f} seconds.")
    print("====================================================")

File: iam-policy-lens/scripts/go/analyzer.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"time"
)

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage:")
		fmt.Println("  go run scripts/go/analyzer.go <path_to_project>")
		os.Exit(1)
	}

	projectPath, err := filepath.Abs(os.Args[1])
	if err != nil {
		fmt.Printf("Error resolving project path: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Scanning: %s for Google Cloud Go API calls\n", projectPath)

	startTime := time.Now()

	calls, err := ScanProject(projectPath)
	if err != nil {
		fmt.Printf("Error scanning project: %v\n", err)
		os.Exit(1)
	}

	elapsed := time.Since(startTime)

	if len(calls) > 0 {
		// Group calls by file path
		grouped := make(map[string][]GapicCall)
		for _, c := range calls {
			grouped[c.FilePath] = append(grouped[c.FilePath], c)
		}

		// Sort file paths for deterministic output
		var files []string
		for f := range grouped {
			files = append(files, f)
		}
		sort.Strings(files)

		for _, f := range files {
			fmt.Printf("\n📄 File: %s\n", f)
			fileCalls := grouped[f]
			sort.Slice(fileCalls, func(i, j int) bool {
				return fileCalls[i].Line < fileCalls[j].Line
			})

			relPath, err := filepath.Rel(projectPath, f)
			if err != nil {
				relPath = f
			}

			for i, c := range fileCalls {
				if i > 0 {
					fmt.Println()
				}
				fmt.Printf("     %s:%d: `%s`\n", relPath, c.Line, c.Source)
				fmt.Printf("     Method: %s [typechecker]\n", c.FullName)
				if c.Credentials != nil {
					fmt.Printf("     Credentials: %s\n", c.Credentials.Source)
					fmt.Printf("     Provenance: CredentialProvenance.%s\n", c.Credentials.Provenance)
					fmt.Printf("     Identity: IdentityContext.%s\n", c.Credentials.Identity)
				}
				permissions := Gapic2Permission(c.FullName)
				if len(permissions) > 0 {
					fmt.Printf("     Permissions: %v\n", permissions)
				}
			}
		}
	} else {
		fmt.Println("No relevant GCP/Google API calls found.")
	}

	fmt.Printf("\nScan completed in %.2f seconds.\n", elapsed.Seconds())
	fmt.Println("====================================================")
}
